packages/contracts/src/contracts/current/utils/LibMem/LibMem.sol

```diff
@@ -0,0 +1,140 @@
+    function getMemAddress(bytes memory input)
+        internal
+        pure
+        returns (uint256 memoryAddress)
+    {
+        assembly {
+            memoryAddress := input
+        }
+        return memoryAddress;
+    }
+
+    /// @dev Copies `length` bytes from memory location `source` to `dest`.
+    /// @param dest memory address to copy bytes to.
+    /// @param source memory address to copy bytes from.
+    /// @param length number of bytes to copy.
+    function memcpy(
+        uint256 dest,
+        uint256 source,
+        uint256 length
+    )
+        internal
+        pure
+    {
+        if (length < 32) {
+            // Handle a partial word by reading destination and masking
+            // off the bits we are interested in.
+            // This correctly handles overlap, zero lengths and source == dest
+            assembly {
+                let mask := sub(exp(256, sub(32, length)), 1)
+                let s := and(mload(source), not(mask))
+                let d := and(mload(dest), mask)
+                mstore(dest, or(s, d))
+            }
+        } else {
+            // Skip the O(length) loop when source == dest.
+            if (source == dest) {
+                return;
+            }
+
+            // For large copies we copy whole words at a time. The final
+            // word is aligned to the end of the range (instead of after the
+            // previous) to handle partial words. So a copy will look like this:
+            //
+            //  ####
+            //      ####
+            //          ####
+            //            ####
+            //
+            // We handle overlap in the source and destination range by
+            // changing the copying direction. This prevents us from
+            // overwriting parts of source that we still need to copy.
+            //
+            // This correctly handles source == dest
+            //
+            if (source > dest) {
+                assembly {
+                    // Record the total number of full words to copy
+                    let nwords := div(length, 32)
+
+                    // We subtract 32 from `send` and `dend` because it
+                    // is easier to compare with in the loop, and these
+                    // are also the addresses we need for copying the
+                    // last bytes.
+                    length := sub(length, 32)
+                    let send := add(source, length)
+                    let dend := add(dest, length)
+
+                    // Remember the last 32 bytes of source
+                    // This needs to be done here and not after the loop
+                    // because we may have overwritten the last bytes in
+                    // source already due to overlap.
+                    let last := mload(send)
+
+                    // Copy whole words front to back
+                    for {let i := 0} lt(i, nwords) {i := add(i, 1)} {
```

Comment:
I don't think we need `i` here, we can loop as long as `source < send` or alternatively `dest < dend`. This saves one `ADD` per loop iteration, plus we don't need `i` and `nwords` anymore.

The explanation of why this works and why it requires `send` and `dend` to be `32` bytes less than `length` is a bit complex, but we can put that in a comment.

Reply:
Yeah I do think that's cleaner. 

The only reason I changed it was to handle the case where `send` turns "negative" and wraps around. We hit this edge case when `0 <= source < 32 - length%32`.

Here's an example of the edge case that you can paste into Remix: https://pastebin.com/fjfh890q

Reply:
I did think of that in my original implementation. This won't happen because nothing is allocated at bytes [0..96) in Solidity. The first two memory slots are scratch space, the third contains the free memory pointer.

Starting from 0.5.0 a fourth will contain a permanent zero, making it 128 bytes. But this zero is also used for `bytes` and `string`s  of length zero, so we might see them in a memcpy.

Reply:
In the `dest > source` case we can also use `lt(dest, dend)`. This will make the edge case much less likely. (It will only happen if both dest and source 1) are longer than 32 bytes, 2) overlap in within 32 bytes and 3) are allocated in the first 32 bytes).

Reply:
We can also use a signed less than `slt` instead of `lt`. Valid addresses will always be less than 2^255, so they can be safely reinterpretted as a signed integer.

```
for {} slt(dest, dend) {} {
```